### DISCUSSION TITLE
Error fileLoaded error when trying to use pyswmm.LidUnit -> posible solution

### DISCUSSION THREAD

**Discussion by antviro** [pyswmm/pyswmm#508] [Q&A] [answered]
**ISSUE+Describe the bug+reproduce**
When I try to initialise a LidUnit I get an error that I think it is caused by line 220 of lidgroups.py that checks model.fileLoaded instead of model._model.fileLoaded (as it is done in other source files within pyswmm). So I think the solution is to add "_model." just before fileLoaded.

Example:

```
In [3]: import pyswmm

In [4]:  sim = pyswmm.Simulation(outINPfile)

 o  Retrieving project data
In [5]: lid_sc = pyswmm.LidUnit(sim, "LS999", "GR")
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [5], in <cell line: 1>()
----> 1 lid_sc = pyswmm.LidUnit(sim, "LS999", "GR")

File ~/.local/lib/python3.9/site-packages/pyswmm/lidgroups.py:220, in LidUnit.__init__(self, model, subcatchmentid, lidid)
    219 def __init__(self, model, subcatchmentid, lidid):
--> 220     if not model.fileLoaded:
    221         raise PYSWMMException("SWMM Model Not Open")
    222     self._model = model

AttributeError: 'Simulation' object has no attribute 'fileLoaded'
```

If I check at the model instance (called here sim) it does not have a fileLoaded attribute but it does have a _model.fileLoaded attribute:

```
In [74]: sim.fileLoaded
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [74], in <cell line: 1>()
----> 1 sim.fileLoaded

AttributeError: 'Simulation' object has no attribute 'fileLoaded'

In [75]: sim._model.fileLoaded
Out[75]: True

```
I think it is a bug, but I am sorry if i were wrong... and thank you in advance for all your efforts!!

I edit to add this:
In my code I have made it work doing 
`sim.fileLoaded = sim._model.fileLoaded`

before calling pyswmm.LidUnit (so I think the suggested solution will make it). And meanwhile if somebody get the error, it can be sorted out just doing so...

```
In [14]: sim.fileLoaded = sim._model.fileLoaded

In [15]: sim.fileLoaded
Out[15]: True

In [16]: sim
Out[16]: <pyswmm.simulation.Simulation at 0x7fa9ae7671c0>

In [17]: lid_sc = pyswmm.LidUnit(sim, "LS999", "GR")

In [18]: 

```

**Desktop (please complete the following information):**
 - OS: Linux
 - Python Version 3.9.2 (running within IPython 8.4.0)
 - PySWMM Version v1.5.1 (and also v1.2.0)

**Additional context**

**Comment by bemcdonnell** [comment]
@antviro, first of all - this is an exceptional bug report and has amazing detail!  Great work! Second of all, I agree that this is a bug.  It is true that `Simulation` has no `.fileLoaded` attribute (only `swmm5.PySWMM` has that).  I can probably look into the code this afternoon.  However, before I dive in, can you confirm inside your RPT file that there are errors?  We could be missing some unit testing coverage there if we didn't see that issue in test.

**Comment by antviro** [comment]
> However, before I dive in, can you confirm inside your RPT file that there are errors? We could be missing some unit testing coverage there if we didn't see that issue in test.

Sorry I don't get what you mean here.

After I did the trick Simulation.fileLoaded = Simulation._model.fileLoaded, it seems to work. Although I think I have more errors beside this, but I think they are not bugs, but that I have not yet learnt how to communicate with LID units [link to stack overflow question](https://stackoverflow.com/questions/77788742/pyswmm-how-to-read-or-set-area-or-number-of-lid-usage).

**Comment by bemcdonnell** [comment, marked as answer]
@antviro I think I found the issue

This is good timing too since I am in the process of overhauling the LID documentation.  It looks like you are trying to directly instantiate an `LidUnit`.  Instead of doing that, use the `LidGroups`.


Here are the edits coming to the documentation. I haven't released these yet but they will be out soon.  Honestly, we have needed to edit these docs for a very long time and I'm sorry it's taken so long!

### LidGroups Iterator Methods.


This allows the user to iterate and get one of the LID Groups as defined inside SWMM's [LID_USAGE] section.  For example, here is a sample section from a SWMM INP file.


```
[LID_USAGE]
;;Subcatchment   LID Process   Number  Area    Width   InitSat  FromImp  ToPerv   RptFile  DrainTo  FromPerv
;;-------------- ------------- ------- ------- ------- -------- -------- -------- -------- -------- ----------
J1_J3_qqqqqqq    LID_C1        1       180.00  2       0        0        0        *        *        0
J11_J14_qqqqqqq  LID_C1        1       140     2       0        0        0        *        *        0
```

Now we can peform manipulations in PySWMM and get objects.

```python
from pyswmm import Simulation, LidGroups

with Simulation('lid_model.inp') as sim:
    for lid_group in LidGroups(sim):
        print(lid_group)

```
```
>>> J1_J3_qqqqqqq
>>> J11_J14_qqqqqqq
```

Once a handle has been created of an LidGroup object we can now make adjustments to the parameters and change the behavior.  The following example provides some context around the different object handles created.  One feature presented herein shows that during a simulation the drain_node (or drain subcatchement) can be changed during the simulation.

```python
from pyswmm import Simulation, LidGroups

with Simulation('lid_model.inp') as sim:
    # This returns an LidGroup object
    lid_group_j1_j3 = LidGroups(sim)["J1_J3_qqqqqqq"]

    # Next to get access to one of the LID units
    # inside this subcatchment it is indexed.

    # This returns an LidUnit object
    lid_unit0=lid_group_j1_j3[0]

    for ind, step in enumerate(sim):
        if ind == 50:
            lid_unit0.drain_node = "J04"
```
Continuing on the previous example, either during the simulation or after it completes, the user has access to lid unit results.  The following code provides details on how to access differne LidUnit results

```python
from pyswmm import Simulation, LidGroups

with Simulation('lid_model.inp') as sim:
    # This returns an LidGroup object
    lid_group_j1_j3 = LidGroups(sim)["J1_J3_qqqqqqq"]

    # Next to get access to one of the LID units
    # inside this subcatchment it is indexed.

    # This returns an LidUnit object
    lid_unit0=lid_group_j1_j3[0]

    for ind, step in enumerate(sim):
        if ind == 50:
            lid_unit0.drain_node = "J04"

    surface = lid_unit0.surface
    pavement = lid_unit0.pavement
    soil = lid_unit0.soil
    storage = lid_unit0.storage
    water_balance = lid_unit0.water_balance
    print(water_balance.inflow)
    print(water_balance.evaporation)
    print(water_balance.infiltration)
    print(water_balance.surface_flow)
    print(water_balance.drain_flow)
    print(water_balance.initial_volume)
    print(water_balance.final_volume)
```

**Comment by antviro** [comment]
Thank you very much... I was rushing home, but I come back to check hehe and following the new documentation instructions everything works fine!

Thank you very much!

**Reply by bemcdonnell** [reply]
@antviro Happy PySWMMing! Glad it worked :)
